### DISCUSSION TITLE
Automatic scaling of the PerspectiveProjection parameters

### DISCUSSION THREAD

**Discussion by whugemann** [ImageMagick/ImageMagick#3585] [Development]
I have just used _-distort PerspectiveProjection_ for the first time, having determined the projection parameters with the help of Fiji's SIFT algorithm. That gives you a 3x3 matrix with the projection parameters, and you have to devide them all by the matrix element a_33 in order to get the normated version (a_33 = 1) that you need for ImageMagick.

It would be reasonable if ImageMagick would automatically perform this normation in case that you provide 9 parameters. I guess, quite a few programs will give you the complete 3x3 matrix without normation.

The situation is somewhat inverse to barrel distortion, where ImageMagick automatically calculates the parameter D by the non-scaling restraint D = 1 - A - B - C.

BTW: There is a typo in the _-verbose_ output for _-distort perspective_; it must be "FX Equivalent".

**Comment by snibgo** [comment]
The numbers in "-distort PerspectiveProjection" can be %[fx:...] expressions, so you can do the division by a_33 there.

**Reply by whugemann** [reply]
Good suggestion. This will at least avoid the use of an external program (like Excel). But the command line will become rather cumbersome.

I however still see the parallels to lens correction / barrel distortion, where a much easier calculation is performed automatically, based on the parameter count.

I admit that the use of _PerspectiveProjection_ is a rather exotic application, so the actual use of the parameters may not be much importance to the normal user.

**Comment by snibgo** [comment]
Allowing nine parameters in "-distort PerspectiveProjection", with automatic division by the last, may be a good idea. The %[fx:...] trick is a possible workaround.

**Reply by fmw42** [reply]
In principle, this is a good idea. It costs little to add to the code and makes no difference if he 9th argument is 1. It also makes it easy to import values from other tools. However, there are lots of little details that would need changing. First, the argument parsing needs one more argument. Second all of the code that describes the -fx equivalent and verbose output would need changing. Then there is the issue of backward compatibility. So I am not sure it is worth the effort.

**Reply by whugemann** [reply]
I'm not suggesting to make it a mandatory parameter: If you provide it, the other 8 parameters are normated by it, if not, the treatment remains as it has been. Thus backward compatibility is preserved.

I am somewhat insecure, whether this is a Git 'issue', i. e. whether I should 'report' this. Do the developers still take a look on this discussion board?

**Reply by fmw42** [reply]
This is a GIT forum and the developers do look at every topic.

**Comment by whugemann** [comment]
I have just stumbled on the same problem again, now using the work-around provided by -fx, which works fine but is rather cumbersome. Afterwards, I wondered whether I already posted this 'problem' over here ...
Seems like my proposal hasn't been set into code yet. (I run the latest version of IM.)
